Source/WebCore/style/CustomPropertyRegistry.cpp

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright (C) 2022 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "CustomPropertyRegistry.h"
+
+#include "CSSPropertyParser.h"
+#include "CSSRegisteredCustomProperty.h"
+#include "StyleBuilder.h"
+#include "StyleResolver.h"
+#include "StyleScope.h"
+
+namespace WebCore {
+namespace Style {
+
+CustomPropertyRegistry::CustomPropertyRegistry(Scope& scope)
+    : m_scope(scope)
+{
+}
+
+const CSSRegisteredCustomProperty* CustomPropertyRegistry::get(const AtomString& name) const
+{
+    // API wins.
+    // https://drafts.css-houdini.org/css-properties-values-api/#determining-registration
+    if (auto* property = m_propertiesFromAPI.get(name))
+        return property;
+
+    return m_propertiesFromStylesheet.get(name);
+}
+
+bool CustomPropertyRegistry::registerFromAPI(CSSRegisteredCustomProperty&& property)
+{
+    // First registration wins.
+    auto result = m_propertiesFromAPI.ensure(property.name, [&] {
```

Comment:
I might call this `isNewEntry` or something like that to be a bit more clear than result.